### PR TITLE
Aperture photometry plugin reflects display unit selection

### PR DESCRIPTION
**Pull request by cshanahan1**
Update all display and outputs of aperture photometry app to reflect selected display unit in cubeviz.
The output photometry table now has a separate column for Unit.

### CODE REVIEW

**Comment by camipacifici** [comment]
Looks good! I just noticed a couple of things:
- with some units the numbers can be very large or very small. Can you please use scientific notation for the min/max/ave/etc in the table? They all come out as zeros with very small/big numbers now.
- at the end of the table, the flux_scaling is not updating to follow the unit change.
- i created a subset and run aperture photometry. the plot is labeled x/y. rerunning it, the plot gets the correct units.
- then I changed units and rerun aperture photometry without changing anything else and the plot kept the points from before and updated only the line
Screenshot 2024-07-31 at 2 50 08 PM
Screenshot 2024-07-31 at 2 50 16 PM
Screenshot 2024-07-31 at 2 50 56 PM

**Comment by camipacifici** [comment]
Just noticed that the problem about not showing the units right away is also in the released version.

**Comment by pllim** [comment]
Now that you mention the radial plot, I wonder how the Gaussian1D fit would behave as well.

**Comment by cshanahan1** [comment]
I tested the gaussian fit and the plotting and it was working for me, but maybe something is getting out of sync. looking into it

**Comment by camipacifici** [comment]
I think the raw profile is not updating and only the label is updating. After changing the units, there is a big difference between the raw and average although they should be comparable.
Screenshot 2024-07-31 at 4 18 36 PM
Screenshot 2024-07-31 at 4 18 44 PM

**Comment by cshanahan1** [comment]
the latest commit should fix the issue with the raw radial profile plot

**Comment by camipacifici** [comment]
The fix worked, thank you!

**Comment by pllim** [comment]
1. Please add a change log.
2. Please fix the test failure in devdeps job.

```
>       row = cubeviz_helper.get_aperture_photometry_results()[0]
E       TypeError: 'NoneType' object is not subscriptable

jdaviz/configs/cubeviz/plugins/tests/test_cubeviz_aperphot.py:102: TypeError
```

**Comment by pllim** [comment]
I still see the error in devdeps job. I suspect the photometry failed somewhere but it was swallowed by try-except block. Please run that test case manually and instead of swallowing the exception, raise it. Then you will see what is really going on. My bet is on either astropy-dev or photutils-dev.

**Comment by cshanahan1** [comment]
still working on figuring out that failure, will update with a fix once i figure it out

File: jdaviz/configs/cubeviz/plugins/tests/test_cubeviz_aperphot.py [hunk 1]
```diff
@@ -192,3 +144,105 @@ def test_cubeviz_aperphot_cube_orig_flux_mjysr(cubeviz_helper, spectrum1d_cube_c
...
+    # after the unit change
+    assert ap.flux_scaling == 3631
+    assert ap.background_value == 4.6e7
+
+    # re-do photometry and make sure table is in new units
+    # and consists of the same results as before converting units
+    ap.vue_do_aper_phot()
+    new_tab = Table(ap.results)
+
+    _compare_table_units(orig_tab, new_tab)
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/tests/test_cubeviz_aperphot.py:226]: If we want to check that new result must be in new unit, "is equivalent" check is not enough. I think it is more correct to check explicitly that old result is in old unit, and new result is in new unit. If new result is still in old unit, old unit is still equivalent to old unit, and the check still passes though it is not really what you want.

File: jdaviz/configs/cubeviz/plugins/tests/test_cubeviz_aperphot.py [hunk 2]
```diff
@@ -192,3 +144,105 @@ def test_cubeviz_aperphot_cube_orig_flux_mjysr(cubeviz_helper, spectrum1d_cube_c
...
+    assert ap.display_flux_or_sb_unit == 'MJy / sr'
+    assert ap.flux_scaling_display_unit == 'MJy'
+
+    # and defaults for inputs are in the correct unit
+    assert ap.flux_scaling == 0.003631
+    assert ap.background_value == 46
+
+    # test output of default curve of growth plot
+    # not sure how to access this from the figure
+    # ap_phot.plot.figure.marks[0].???
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/tests/test_cubeviz_aperphot.py:207]: I think you can access the marks here? Kyle might know for sure. You will have to play around with the API to find the correct label, etc.

https://github.com/spacetelescope/jdaviz/blob/f4a68899661d5fd115ce47af78395adf2c6f2458/jdaviz/core/template_mixin.py#L4856

File: jdaviz/configs/cubeviz/plugins/tests/test_cubeviz_aperphot.py [hunk 3]
```diff
@@ -192,3 +144,105 @@ def test_cubeviz_aperphot_cube_orig_flux_mjysr(cubeviz_helper, spectrum1d_cube_c
...
+
+    # check that initial units are synced between plugins
+    assert uc.flux_unit.selected == 'MJy'
+    assert uc.angle_unit.selected == 'sr'
+    assert ap.display_flux_or_sb_unit == 'MJy / sr'
+    assert ap.flux_scaling_display_unit == 'MJy'
+
+    # and defaults for inputs are in the correct unit
+    assert ap.flux_scaling == 0.003631
+    assert ap.background_value == 46
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/tests/test_cubeviz_aperphot.py:203]: Please use `assert_allclose` for floating point checks.

Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/tests/test_cubeviz_aperphot.py:203]: Will the results here change when https://github.com/spacetelescope/jdaviz/pull/3133 is merged? 🤪

File: jdaviz/configs/cubeviz/plugins/tests/test_cubeviz_aperphot.py (hunk 2 above)
Comment (kecnry) [reply jdaviz/configs/cubeviz/plugins/tests/test_cubeviz_aperphot.py:207]: since these were (I think from a brief skim of the code) added as "data-collection" entries to the plugin plot, you should be able to access them via `ap_phot.plot.layers`.  `ap_phot.plot.marks` are individual bqplot marks that are not controlled by the data-collection.

**Comment by cshanahan1** [comment]
a few things:

1. There is no current test coverage for batch mode in cubeviz. I have run into occasional errors testing the unit conversion with batch mode, but also sometimes it does work. I started working on adding some test coverage for batch mode specifically (both generally and testing before/after unit conversion) but there are some existing issues making this complicated. For example, an error is thrown when you try to toggle on batch mode by setting plugin.multiselect to True (which works in imviz), which then doesn't allow you to select multiple apertures or use the 'unpack_batch_options' method. I think that validating that batch mode works and is covered in tests warrants its own ticket outside of this work, and that not testing batch mode + unit conversion shouldn't hold up merging this PR.

2. Similarly, there is no current test coverage for radial profile or curve of growth plots in the plugin. I looked at the markers plugin to get the syntax to access points directly from a figure to test what is happening in the UI (rather than just calling _curve_of_growth, which wouldn't pick up things in the plugin like unit changes, and is already tested on its own in the imviz tests), and that is not working here. Even if I can get it to work, I think that this goes out of scope since those basic tests don't exist already. I propose addressing this in the same followup that adds test coverage for batch mode, and adding a test there that covers unit conversion.

File: jdaviz/configs/cubeviz/plugins/tests/test_cubeviz_aperphot.py (hunk 1 above)
Comment (cshanahan1) [reply jdaviz/configs/cubeviz/plugins/tests/test_cubeviz_aperphot.py:226]: done, added a check that the table units are what they should be so it doesn't accidentally pass if they match

File: jdaviz/configs/cubeviz/plugins/tests/test_cubeviz_aperphot.py (hunk 2 above)
Comment (cshanahan1) [reply jdaviz/configs/cubeviz/plugins/tests/test_cubeviz_aperphot.py]: i made a comment below - i think that since test coverage doesn't already exist for the plugin plots, this should be added as a follow up.

**Comment by pllim** [comment]
I thought we agree to disallow magnitude in aperture photometry? This is probably wrong and should not be allowed.

Screenshot 2024-08-06 153338

**Comment by pllim** [comment]
1. I don't know what the flux scaling mean in unit that isn't a variant of Jy. The original ask from user was to hardcode 3631 Jy (ref: https://en.wikipedia.org/wiki/AB_magnitude) as a convenient shortcut to quickly get AB mag for JWST data. Maybe it is cleaner to set it to 1 for something that isn't Jy (or some prefix of it)?
2. Not sure what is going on with the Curve of Growth JSON message.

**Comment by pllim** [comment]
@kecnry also said we should disable multi-select for aperture photometry in Cubeviz. Is that part of this PR or a different ticket?

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py [hunk 4]
```diff
@@ -953,6 +1177,18 @@ def _curve_of_growth(data, centroid, aperture, final_sum, wcs=None, background=0
...
+    # cubeviz only to handle unit conversion display unit changes
+    if display_unit is not None:
+        sum_unit = u.Unit(display_unit)
+    else:
+        if isinstance(data, u.Quantity):
+            sum_unit = data.unit
+        else:
+            sum_unit = None
+    if sum_unit and pixarea_fac is not None:
+        sum_unit *= pixarea_fac.unit
```
Comment (pllim) [inline jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:1190]: I don't understand what this is doing. As discussed, in `main`, JWST data with MJy/sr naturally have the `/sr` removed during curve of growth calculation. Theoretically, this function should not have to change at all in this PR. Why is this change needed again?

File: jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py [hunk 5]
```diff
@@ -547,13 +550,12 @@ def extract(self, return_bg=False, add_data=True, **kwargs):
         spec.meta['_pixel_scale_factor'] = pix_scale_factor
 
         # inform the user if scale factor keyword not in metadata
-        if 'PIXAR_SR' not in self.cube.meta:
-            snackbar_message = SnackbarMessage(
+        if _first_run and ('PIXAR_SR' not in self.cube.meta):
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py:553]: I fixed the snackbar spam here because it is too annoying to review with the bug in this PR.

This is a "quick and dirty" fix. Is it worrisome that this line is called so many times? Probably but not something I want to fix as a reviewer of this PR.

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py (hunk 4 above)
Comment (cshanahan1) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:1190]: the initial sum_unit is always the data unit (MJy / sr), and if it is multiplied by pixarea_fac (which is how it is removed for the final sum / curve of growth) it will be MJy. This block handles translating the output sum units to the display unit if they differ

**Comment by kecnry** [comment]
> @kecnry also said we should disable multi-select for aperture photometry in Cubeviz. Is that part of this PR or a different ticket?

This is already a separate ticket, and I just said disabling could be an option if it isn't an easy fix (but that we probably should try to get it working and covered in tests if there aren't too many roadblocks).

File: jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py (hunk 5 above)
Comment (kecnry) [reply jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py:553]: We have a plan to address this as well, so will need to revert this full commit, is it needed to have until then?

Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py:553]: I think this temporary fix should go in because otherwise the snackbar is drowned out and I cannot see the actually useful  messages.

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.vue [hunk 6]
```diff
@@ -99,7 +100,7 @@
             label="Pixel area"
             v-model.number="pixel_area"
             type="number"
-            hint="Pixel area in arcsec squared, only used if sr in data unit"
+            hint="Pixel area in arcsec squared, only used data is in surface brightness."
```
Comment (bmorris3) [inline jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.vue:103]:
```suggestion
            hint="Pixel area in arcsec squared, only used if data is in units of surface brightness."

```

**Review by bmorris3 — COMMENTED** [review]
First comment from a first glance. More comments tomorrow!

File: jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py (hunk 5 above)
Comment (cshanahan1) [reply jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py:553]: i dont know if it makes sense to have the snackbar only pop up when you open the extraction plugin, since an extraction is done before that, but if this is temporary then thats fine

Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py:553]: Another option is to just temporary comment this out entirely until someone can fix it properly. (I actually think it should only emit once at load time, not here.) As it is right now, it is doing more harm than good.

Comment (kecnry) [reply jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py:553]: this doesn't seem to be too related to the scope of this PR, but if its absolutely necessary to get rid of here instead of in a separate PR, then I would vote to have it just removed (since that is the eventual end result) rather than introducing new code that we know we'll be getting rid of in an upcoming effort

Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py:553]: Another question is why is this code invoked so many times? Is it oversubscribing to events? Are we doing unnecessary re-calculations? But that is, again, out of scope here.

Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py:553]: If this really bothers you, I can undo these changes (or Clare can). I can open up a separate PR.

Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py]: OK I undid the changes here. I think Kyle said he is going to look into it separately.

**Comment by pllim** [comment]
As it is, `background` and `flux_scaling` is going to crash whenever you try to convert from per-wavelength to per-frequency because that conversion needs `u.spectral_density` that needs a wavelength/frequency as input. Example traceback:

```
File aper_phot_simple.py:184, in SimpleAperturePhotometry._on_display_units_changed(self, event)
    181     new_unit = u.Unit(self.display_flux_or_sb_unit)
    183     bg = self.background_value * prev_unit
--> 184     self.background_value = bg.to_value(new_unit)
    186 # convert flux scaling to new unit
    187 if self.flux_scaling is not None:

UnitConversionError: 'MJy / sr' (surface brightness) and 'erg / (Angstrom s sr cm2)' (surface brightness wav) are not convertible
```

For Cubeviz, you can probably grab this value from slice. But this is going to be much trickier if you ever want to enable the same functionality for Imviz because then you need to infer that value from some bandpass information that might or might not be available.

**Comment by cshanahan1** [comment]
oh, i was working on adding the equivalencies but it looks like you just committed those changes?

**Review by pllim — APPROVED** [review]
I think it is okay to not convert biweight midvariance when it fails (it will if you convert per-wave to per-freq because it is flux^2). I don't think anyone is going to notice. Good enough for now.

**Comment by pllim** [comment]
Re: https://github.com/spacetelescope/jdaviz/pull/3118#issuecomment-2278044827

Oh ooops... Feel free to drop my commits then if you want to push yours out.

**Comment by pllim** [comment]
Noooo... I think I broke it.

**Comment by pllim** [comment]
Actually maybe the difference is caused by https://github.com/spacetelescope/jdaviz/pull/3133

**Review by bmorris3 — APPROVED** [review]
This looks good to me, thanks for your perseverance @cshanahan1. It's unfortunate that the units astronomers love are so messy, but I think this makes it clearer what we're doing with them.
